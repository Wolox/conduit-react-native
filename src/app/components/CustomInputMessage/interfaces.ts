import { ReactNode } from 'react';
import { ImageSourcePropType, ViewStyle } from 'react-native';
import { ControllerProps } from 'react-hook-form';

type Controller = ControllerProps;

export interface CustomInputMessage {
  messageLabel?: string;
  maxLengthMessage: number;
  minLengthMessage: number;
  messageRules?: Controller['rules'];
  numberOfLinesMessage?: number;
  keyboardAwareViewChildren?: ReactNode;
  placeHolder?: string;
  showPlaceholderMessage?: boolean;
  onChangeEventMessage?: (message: string) => void;
  automaticScroll?: boolean;
  styleInputText: ViewStyle;
  messageButton?: string;
  onPressButton?: () => void;
  iconButton?: ImageSourcePropType;
}
