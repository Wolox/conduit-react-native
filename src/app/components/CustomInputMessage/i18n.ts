import i18next from 'i18next';

i18next.addResources('es', 'FEEDBACK_CUSTOM_INPUT_MESSAGE', {
  MESSAGE_LABEL: 'Env√≠anos tu comentario',
  OPTIONAL: ' (Opcional)',
  MESSAGE_PLACEHOLDER: 'Ingresar tu comentario...',
  MAX_MESSAGE: 'Hasta {{max}} caracteres'
});
