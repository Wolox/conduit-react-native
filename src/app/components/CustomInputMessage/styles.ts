import { StyleSheet } from 'react-native';
import { black } from '@constants/colors';
import { IS_SMALL_DEVICE } from '@constants/platform';
import { SPACING } from '@constants/dimentions';

const WIDTH_LABEL = IS_SMALL_DEVICE ? '90%' : '100%';

const COMMON_PADDING_INPUT: number = 8;

export default StyleSheet.create({
  container: {
    marginHorizontal: SPACING.XXXSMALL
  },
  inputContainer: {
    marginTop: SPACING.XXSMALL
  },
  messageLabel: {
    paddingTop: 24,
    justifyContent: 'space-evenly',
    width: WIDTH_LABEL,
    flexDirection: 'row'
  },
  customLabel: {
    marginRight: SPACING.NONE
  },
  textInputMessage: {
    fontSize: 14,
    color: black,
    paddingVertical: SPACING.NONE,
    textAlignVertical: 'top',
    marginBottom: 20,
    paddingHorizontal: COMMON_PADDING_INPUT,
    paddingTop: COMMON_PADDING_INPUT
  },
  customTextInput: {
    paddingBottom: SPACING.NONE,
    marginBottom: SPACING.NONE
  },
  customInputContainer: {
    height: 300,
    paddingBottom: 10
  },
  error: {
    marginLeft: SPACING.MEDIUM,
    width: '80%'
  }
});
