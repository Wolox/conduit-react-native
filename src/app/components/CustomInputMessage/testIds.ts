const testIds = {
  messageFeedback: 'message-feedback'
};

export default testIds;
