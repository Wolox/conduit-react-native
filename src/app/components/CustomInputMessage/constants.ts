import { DeepMap, FieldError } from 'react-hook-form';
import './i18n';

export const FIELDS = {
  message: 'message'
} as const;

export interface CustomInputMessageInterface {
  [FIELDS.message]: string;
}
export interface TextInputKeyEvent {
  nativeEvent: {
    key: string;
  };
}

export const validations = (errors: DeepMap<CustomInputMessageInterface, FieldError>) => ({
  ERRORS_MESSAGE: (errors?.message?.message && errors?.message?.type === 'pattern') || false,
  MESSAGE_HAS_ERRORS_MESSAGE: !!errors?.message?.message
});
