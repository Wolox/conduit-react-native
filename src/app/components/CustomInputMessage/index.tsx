import React, { memo, useRef, useState } from 'react';
import { TextInput, View } from 'react-native';
import { useForm } from 'react-hook-form';
// import i18next from 'i18next';
import CustomText from '@components/CustomText';
import ControlledCustomTextInput from '@components/CustomTextInput/controller';
import { Nullable } from '@interfaces/globalInterfaces';
// import { getAccessibilityLabel } from '@utils/accessibilityUtils';
// import { KeyboardAwareScrollView } from 'react-native-keyboard-aware-scroll-view';
import { isIos } from '@constants/platform';

import { FIELDS, validations, TextInputKeyEvent, CustomInputMessageInterface } from './constants';
import { CustomInputMessage as Props } from './interfaces';
import testIds from './testIds';
import './i18n';
import styles from './styles';

function CustomInputMessage({
  messageLabel,
  maxLengthMessage,
  numberOfLinesMessage,
  keyboardAwareViewChildren,
  messageRules,
  showPlaceholderMessage,
  onChangeEventMessage,
  styleInputText
}: Props) {
  const {
    trigger,
    control,
    getValues,
    formState: { errors }
  } = useForm<CustomInputMessageInterface>({
    mode: 'all',
    reValidateMode: 'onChange',
    shouldFocusError: true
  });
  const messageInputRef = useRef<Nullable<TextInput>>(null);
  const [showErrorMessage, setShowErrorMessage] = useState(false);
  const [lastMessageLength, setLastMessageLength] = useState(0);
  const { ERRORS_MESSAGE, MESSAGE_HAS_ERRORS_MESSAGE } = validations(errors);
  const [currentFocus, setCurrentFocus] = useState('');
  const [placeHolderRightMessage, setPlaceHolderRightMessage] = useState(`0/${maxLengthMessage}`);
  const { message } = getValues();
  const onBlur = (field: string) => {
    setCurrentFocus('');
    const fieldName = FIELDS[field as keyof typeof FIELDS];
    setShowErrorMessage(false);
    trigger(fieldName);
  };
  const onFocus = (field: string) => {
    const setField = FIELDS[field as keyof typeof FIELDS];
    setCurrentFocus(setField);
  };

  const onChangeTextMessage = (value: string) => {
    const cleanValue = value.trim();
    setShowErrorMessage(false);
    setPlaceHolderRightMessage(`${cleanValue.length}/${maxLengthMessage}`);
    trigger();
    if (onChangeEventMessage) onChangeEventMessage(value);
  };

  const isFocusMessage = currentFocus === FIELDS.message;
  const onKeyPressMessage = ({ nativeEvent: { key } }: TextInputKeyEvent) => {
    const { message: newMessage } = getValues();
    const isBackSpaceKey = key === 'Backspace';
    const sizeKey = isBackSpaceKey ? 0 : 1;
    const lastKeySize = message.trim().length + sizeKey;
    setLastMessageLength(lastKeySize);
    if (lastMessageLength > maxLengthMessage && !isIos) setShowErrorMessage(true);
    if (isIos && newMessage.length === maxLengthMessage) setShowErrorMessage(true);
    if (isBackSpaceKey) setShowErrorMessage(false);
  };

  return (
    <View style={styles.container}>
      {/* <KeyboardAwareScrollView
        bounces={false}
        enableOnAndroid
        // style={{ flex: 1, height: '100%' }}
        contentContainerStyle={{ flex: 1, backgroundColor: 'red', heigth: '100%' }}
        > */}
      {keyboardAwareViewChildren && keyboardAwareViewChildren}
      <View style={styles.messageLabel}>
        <CustomText
          style={styles.customLabel}
          darkGray
          darkViolet={
            (currentFocus === FIELDS.message && !showErrorMessage && !MESSAGE_HAS_ERRORS_MESSAGE,
            MESSAGE_HAS_ERRORS_MESSAGE)
          }
          error={showErrorMessage || ERRORS_MESSAGE}
          medium>
          {messageLabel}
        </CustomText>
        {/* <CustomText poppins small gray>
            {'opcional'}
          </CustomText> */}
      </View>
      <View style={styles.inputContainer}>
        <ControlledCustomTextInput
          inputContainerStyle={[styles.customInputContainer, styleInputText && styleInputText]}
          ref={messageInputRef}
          testID={testIds.messageFeedback}
          control={control}
          name={FIELDS.message}
          placeholder={'deja tu comentario'}
          onChange={onChangeTextMessage}
          onFocus={() => onFocus(FIELDS.message)}
          onBlur={() => onBlur(FIELDS.message)}
          placeholderRightText={placeHolderRightMessage}
          inputTextStyles={styles.textInputMessage}
          maxLength={maxLengthMessage}
          numberOfLines={numberOfLinesMessage}
          multiline
          blurOnSubmit
          showPlaceholderRight={isFocusMessage || showPlaceholderMessage}
          // showPlaceholderRight={true}
          error={showErrorMessage || MESSAGE_HAS_ERRORS_MESSAGE}
          showIconError={showErrorMessage || MESSAGE_HAS_ERRORS_MESSAGE}
          onKeyPress={onKeyPressMessage}
          rules={messageRules}
          style={styles.customTextInput}
          // accessibilityLabel={getAccessibilityLabel(
          //   testIds.messageFeedback,
          //   i18next.t('FEEDBACK_CUSTOM_INPUT_MESSAGE:MESSAGE_LABEL')
          // )}
        />
        {showErrorMessage && (
          <CustomText error poppinsSemiBold xsmall style={styles.error}>
            {`cantida de caracteres ${maxLengthMessage}`}
          </CustomText>
        )}
        {errors?.message?.message && isFocusMessage && !showErrorMessage && (
          <CustomText error poppinsSemiBold xsmall style={styles.error}>
            {errors.message.message}
          </CustomText>
        )}
      </View>
      {/* </KeyboardAwareScrollView> */}
    </View>
  );
}

CustomInputMessage.defaultProps = {
  showButtons: true,
  automaticScroll: true
};

export default memo(CustomInputMessage);
