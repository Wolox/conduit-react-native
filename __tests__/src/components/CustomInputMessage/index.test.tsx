import React from 'react';
import renderer from 'react-test-renderer';
import CustomInputMessage from '@components/CustomInputMessage';

describe('CustomInputMessage', () => {
  test('CustomInputMessage Snapshot', () => {
    const item = {
      maxLengthMessage: 5,
      minLengthMessage: 0,
      styleInputText: {}
    };
    const tree = renderer.create(<CustomInputMessage {...item} />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
