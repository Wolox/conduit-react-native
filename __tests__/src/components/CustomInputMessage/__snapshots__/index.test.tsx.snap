// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CustomInputMessage CustomInputMessage Snapshot 1`] = `
<View
  style={
    Object {
      "marginHorizontal": 2,
    }
  }
>
  <View
    style={
      Object {
        "marginTop": 4,
      }
    }
  >
    <View
      style={
        Array [
          Object {
            "marginBottom": 5,
            "padding": 10,
          },
          false,
          Object {
            "marginBottom": 0,
            "paddingBottom": 0,
          },
        ]
      }
    >
      <View
        style={
          Array [
            Object {
              "backgroundColor": "#FFF",
              "borderRadius": 15,
              "elevation": 1,
              "height": 75,
              "paddingHorizontal": 5,
              "shadowColor": "#E5E5E5",
              "shadowOffset": Object {
                "height": 3,
                "width": 0,
              },
              "shadowOpacity": 1,
            },
            Object {},
            Array [
              Object {
                "height": 300,
                "paddingBottom": 10,
              },
              Object {},
            ],
            Array [
              Object {
                "height": 300,
                "paddingBottom": 10,
              },
              Object {},
            ],
          ]
        }
      >
        <TextInput
          allowFontScaling={false}
          autoCapitalize="none"
          autoCompleteType="off"
          autoCorrect={false}
          blurOnSubmit={true}
          clearButtonMode="never"
          editable={true}
          keyboardType="default"
          maxLength={5}
          multiline={true}
          name="message"
          onBlur={[Function]}
          onChangeText={[Function]}
          onFocus={[Function]}
          onKeyPress={[Function]}
          placeholder=""
          rejectResponderTermination={true}
          returnKeyType="done"
          style={
            Array [
              Object {
                "color": "#555555",
                "fontFamily": "Nunito",
                "fontSize": 14,
                "fontStyle": "normal",
                "fontWeight": "400",
                "margin": 0,
                "padding": 0,
              },
              false,
              Object {
                "color": "#000",
                "fontSize": 14,
                "marginBottom": 20,
                "paddingHorizontal": 8,
                "paddingTop": 8,
                "paddingVertical": 0,
                "textAlignVertical": "top",
              },
              Object {
                "color": "#000",
                "fontSize": 14,
                "marginBottom": 20,
                "paddingHorizontal": 8,
                "paddingTop": 8,
                "paddingVertical": 0,
                "textAlignVertical": "top",
              },
            ]
          }
          underlineColorAndroid="transparent"
          value=""
        />
      </View>
      <View
        style={
          Array [
            Object {
              "height": 15,
              "marginTop": 3,
            },
            undefined,
          ]
        }
      />
    </View>
  </View>
</View>
`;
